src/v/kafka/server/handlers/fetch.cc

```diff
@@ -1009,20 +1198,36 @@ class fetch_worker {
  */
 class nonpolling_fetch_plan_executor final : public fetch_plan_executor::impl {
 public:
-    explicit nonpolling_fetch_plan_executor(bool debounce = false)
+    nonpolling_fetch_plan_executor()
       : _last_result_size(ss::smp::count, 0)
-      , _debounce(debounce)
       , _fetch_timeout{[this] { _has_progress.signal(); }} {}
 
     /**
      * Executes the supplied `plan` until `octx.should_stop_fetch` returns true.
      */
     ss::future<> execute_plan(op_context& octx, fetch_plan plan) final {
-        if (_debounce) {
+        const auto& fetch_sg
+          = octx.rctx.server().local().fetch_scheduling_group();
+
+        auto fetch_read_strategy
+          = config::shard_local_cfg().fetch_read_strategy();
+        if (
+          fetch_read_strategy
+          == model::fetch_read_strategy::non_polling_with_debounce) {
             co_await ss::sleep(std::min(
               config::shard_local_cfg().fetch_reads_debounce_timeout(),
               octx.request.data.max_wait_ms));
         }
+        // Ensure both fetch debounce and the fetch scheduling group are enabled
+        // before trying to apply any delay.
+        if (
+          fetch_read_strategy
+            == model::fetch_read_strategy::non_polling_with_pid
+          && config::shard_local_cfg().use_fetch_scheduler_group()) {
+            static thread_local pid_controller pid = {};
```

Comment:
Why is this a `static thead_local`? It is an unusual lifetime. Shouldn't it have the same lifetime as the kafka server?

Reply:
It was mainly to keep the implementation contained in the fetch handler. Otherwise I'd have to break out the pid controller into its own .h/.cc file. Maybe its not a bad idea though.

Reply:
Is it right and/or intended for two kafka server objects which exist simultaneously on the same shard to share the controller? Normally I would expect each server to have independent state. Same for two servers which do not exist simultaneously but are created then destroyed one after the other.

Reply:
I guess it'd be possible to have different kafka server objects on the same thread. I'll create a different pid controller per server object to be safe.